Create n8n.mdx (#1358)
ngrok/ngrok-docs: docs/guides/other-guides/n8n.mdx

---
title: Using ngrok with n8n to make your workflow public 
description: Learn how to use ngrok to make your n8n workflow accessible from anywhere, not just your local host. 
---

import TabItem from "@theme/TabItem";
import Tabs from "@theme/Tabs";

[n8n allows you to self-host](https://docs.n8n.io/hosting/?) which is great when you're testing workflows, but when you're ready to turn your instance into a production-ready automation and don't want to buy a domain or pay for hosting, you might feel lost. This tutorial walks you through how to get your self-hosted n8n instance running in Docker live on the internet. We'll show you how to set everything up in Docker, get your free static domain, and how to protect your endpoint, all for free with ngrok.

## Set up your local server with n8n

Follow the steps listed in [n8n's docs](https://docs.n8n.io/hosting/installation/docker/) to self-host n8n in Docker.

### 1. Install ngrok

[Download and install](https://ngrok.com/download) the ngrok CLI.

**For Docker**
Install ngrok via Docker with the following command:

```
docker pull ngrok/ngrok
```

Run ngrok via Docker:

```

> docker run -it -e NGROK_AUTHTOKEN=<token> ngrok/ngrok http 80

Reviewer: Flow here seems out of order. This mentions the auth token but next step is obtaining it.

```

### 2. Connect ngrok to your account

Sign up for an account at [ngrok.com](https://ngrok.com/) and run:

```bash

> ngrok config add-authtoken YOUR_NGROK_AUTH_TOKEN

Reviewer: This would be the agent outside of docker in this case?

Author: ah yes good catch

```

### 3. Claim your free static domain

Head over to https://dashboard.ngrok.com/domains to claim your free static domain. It should look something like happy-rabbit-partially.ngrok.free.app. Remember it, as that's your free ngrok domain.

Copy this domain to use in the next step.

### 4. Launch a new container in Docker with updated environment variables including your ngrok domain

Create a new container in Docker.

> Important: Reuse the same volume or bind mount for the data folder in your old n8n container. The host path must stay the same. This will ensure that your n8n workflows,credentials, or execution history are not impacted.

Reviewer: ```suggestion
Important: Reuse the same volume or bind mount for the data folder in your old n8n container. The host path must stay the same. This will ensure that your n8n workflows, credentials, or execution history are not impacted.
```

---
sidebar_label: Using ngrok with n8n
title: Using ngrok with n8n to make your workflow public 
description: Learn how to use ngrok to make your n8n workflow accessible from anywhere, not just your local host. 
---

import TabItem from "@theme/TabItem";
import Tabs from "@theme/Tabs";

> [n8n allows you to self-host](https://docs.n8n.io/hosting/?) which is great when you're testing workflows, but when you're ready to turn your instance into a production-ready automation and don't want to buy a domain or pay for hosting, you might feel lost. This tutorial walks you through how to get your self-hosted n8n instance running in Docker live on the internet. We'll show you how to set everything up in Docker, get your free static domain, and how to protect your endpoint, all for free with ngrok.

Reviewer 2: Just a few copy tweaks here for clarity.

```suggestion
[n8n allows you to self-host](https://docs.n8n.io/hosting/) which is great when you're testing workflows, but when you're ready to turn your instance into a production-ready automation and don't want to buy a domain or pay for hosting, you might feel lost. 

This guide walks you through how to get your self-hosted n8n instance running in Docker live on the internet. We'll show you how to set ngrok with Docker or or with the agent CLI, get your free static domain, and protect your endpoint, all for free with ngrok.
```

## Set up your local server with n8n

Follow the steps listed in [n8n's docs](https://docs.n8n.io/hosting/installation/docker/) to self-host n8n in Docker.

### 1. Connect ngrok to your account

Sign up for an account at [ngrok.com](https://ngrok.com/) and [get your auth token](https://dashboard.ngrok.com/get-started/your-authtoken).

### 2. Install ngrok

<Tabs groupId="user-type" queryString="user-type">
<TabItem value="Docker" label="Docker" default>

**For Docker**
Install ngrok via Docker with the following command:

```bash
docker pull ngrok/ngrok
```

Run ngrok via Docker:

```bash
docker run -it -e NGROK_AUTHTOKEN=<token> ngrok/ngrok http 80
```

</TabItem>
<TabItem value="CLI" label="CLI">
[Download and install](https://ngrok.com/download) the ngrok CLI. Then connect ngrok to your account by running the following command in your terminal:

```bash
ngrok config add-authtoken YOUR_NGROK_AUTH_TOKEN
```

</TabItem>
</Tabs>

### 3. Claim your free static domain

> Head over to https://dashboard.ngrok.com/domains to claim your free static domain. It should look something like happy-rabbit-partially.ngrok.free.app. Remember it, as that's your free ngrok domain.

Reviewer 2: ```suggestion
Head over to https://dashboard.ngrok.com/domains to claim your free static domain. It should look something like `happy-rabbit-partially.ngrok.free.app`.
```

Reviewer 2: Feels a bit strange to me to say to remember it, but then also copy it in the next paragraph.

Copy this domain to use in the next step.

### 4. Launch a new container in Docker with updated environment variables including your ngrok domain

Create a new container in Docker.

Important: Reuse the same volume or bind mount for the data folder in your old n8n container. The host path must stay the same. This will ensure that your n8n workflows, credentials, or execution history are not impacted.

> Create a new container, select port 5555.

Reviewer 2: Two things here:

- Not sure why this references port 5555 when the n8n docs tell you to put it on 5678.
- This seems to talk about using Docker Desktop, but the docs have you on the CLI.

https://docs.n8n.io/hosting/installation/docker/

Reviewer 3: How should we address this? I'm a little confused because that page seems to reference Docker Desktop. However, I'm open to making any changes you think make sense, as I'm not the original writer of this PR and I'm not entirely sure what to do

> Create an enviroment variable of
> 
> ```bash
> export N8N_EDITOR_BASE_URL=<your_static_domain_here>
> ```
> 
> and paste in your ngrok URL as the value. Make sure it is https.
> 
> Next, create another enviroment variable of
> 
> ```bash
> export WEBHOOK_URL=<your_webhook_url_here>
> ```
> 
> and paste in your ngrok URL as the value. Make sure it is https.

Reviewer 2: I feel like we could simplify this section:

````suggestion
Create two environment variables, pasting in your ngrok URL as the value, making sure it starts with `https://`.

```bash
export N8N_EDITOR_BASE_URL=<your_static_domain_here>
export WEBHOOK_URL=<your_webhook_url_here>
```
````

Lastly, add the following environment variable:

```bash
export N8N-DEFAULT_BINARY_DATA_MODE=filesystem
```

When you and set the value to filesystem, n8n saves large files as binary data during workflow runs so your processes run quickly.

Once this is done and the new container with your ngrok domain is launched, delete the old one and hit the run button to create your new container running on ngrok.

### 5. Test your integration

> Open your localhost and go to port `5555`. Log in with your credentials. You should see everything you had locally.

Reviewer 2: Just to flag that if we change the port above, do it here too.

### 6. Expose your server with ngrok

Start a tunnel to your app's port:

```bash

> ngrok http 5555 --url https://<your_static_domain_here>

Reviewer 2: Port!

```

Navigate to your ngrok domain. You should be able to access it from anywhere now.

## Deny traffic from malicious IPs

> Let's limit access to your n8n server from malicious IPs by utilizing ngrok's [Traffic Policy](https://ngrok.com/docs/traffic-policy/) and [IP Intel](https://ngrok.com/docs/traffic-policy/variables/ip-intel/_=)

Reviewer 2: ```suggestion
Let's limit access to your n8n server from IPs associated with Tor exit nodes by utilizing ngrok's [Traffic Policy](/docs/traffic-policy/) and [IP Intel](/docs/traffic-policy/variables/ip-intel/).
```

### 1. Create a Traffic Policy

Locally, save this file as `policy.yml`:

```bash
on_http_request:
  - expressions:
    - "('proxy.anonymous.tor' in conn.client_ip.categories)"
  - actions:
      - type: deny
```

### 2. Restart the agent

> Then, restart your ngrok agent, but this time use:

Reviewer 2: ```suggestion
Restart your ngrok agent with the `--traffic-policy-file` argument.
```
